### DISCUSSION TITLE
Better visualization of Inverter Device Fault states

### DISCUSSION THREAD

**Discussion by davidrapan** [davidrapan/ha-solarman#25] [Ideas]
Hey guys, would like to ask you for some knowledge gathering help with some different states (and what did they proly meant) you observed with this entity which usually show state as:
```
['0x0', '0x0', '0x0', '0x0']
```

As the official docs don't say much about the possible values. So you can follow with your observations here.

Thanks!

**Comment by davidrapan** [comment]
I'll start with one observed but undocumented value (as an example) for SG04LP3:
```
 ['0x0', '0x4', '0x0', '0x0'] - F19 Tz_Integ_Fault
```

**Comment by cyh0928** [comment]
This is too much
['0x0', '0x0', '0x0', '0x0', '0x3243']
['0x0', '0x0', '0x0', '0x0', '0x3242']
['0x10', '0x0', '0x0', '0x0', '0x3142']
['0x0', '0x0', '0x0', '0x0', '0x3124'].。。。。。。。。。

**Reply by davidrapan** [reply]
Any idea of what they could mean? At least some of them?

**Reply by cyh0928** [reply]
According to the information you provided, you have a 16-bit value 0x3100, and I hope to understand the meaning of each bit. This value is expressed in hexadecimal, and converted to binary is 0011000100000000.

Now let's explain it bit by bit:

Bit 0 (Discharge over current): 0, indicating normal.

Bit 1 (Over Load): 0, indicating normal.

Bit 2 (Battery not connected - Bat Didconnect): 1, indicating abnormal.

Bit 3 (Battery capacity is insufficient - Bat Under Capacity): 1, indicating abnormal.

Bit 4 (Battery capacity is low - Bat Low Capacity): 0, indicating normal.

Bit 5 (Battery voltage is high - Bat Over Volt): 0, indicating normal.

Bit 6 (Grid voltage is low - Grid Low Volt): 1, indicating abnormal.

Bit 7 (Grid voltage is high - Grid Over Volt): 0, indicating normal.
Bit 8 (Grid Low Freq): 0, indicating normal.
Bit 9 (Grid Over Freq): 0, indicating normal.
Bit 10 (Gfci Over): 0, indicating normal.
Bit 11 (Parallel CAN communication failure - ParallelCANFail): 0, indicating normal.
Bit 12 (Grid CT Reverse): 0, indicating normal.
Bit 13 (BUS Under Volt): 0, indicating normal.
Bit 14 (BUS Over Volt): 0, indicating normal.
Bit 15 (INV Over Current): 0, indicating normal.
According to the above analysis, there are several abnormal conditions in your 16-bit value 0x3100:

Battery not connected (Bit 2)
Battery capacity is insufficient (Bit 3)
Grid voltage is low (Bit 6)
Other conditions are normal.

**Reply by davidrapan** [reply]
Cool! Thanks!
